alf/experience_replayers/replay_buffer.py

```
@@ -622,26 +654,74 @@ def get_field(self, field_name, env_ids, positions):
+           experience of the next iteration which guarantees its
+           participation in training.
+
+        2. If ``keep_as_earliest_frames`` is set to ``True``, a number
+           of ``num_earliest_frames_ignored`` experiences from the end
+           of the buffer (w.r.t. pos) will be kept in addition to what
+           has been kept by ``keep_last_exp``. This is specifically
+           designed for ``FrameStacker`` so that after the buffer is
+           populated again the first few experiences can still be
+           stacked using the experiences from last round before the
+           last clear() is called.
 
         Args:
 
             keep_last_exp: see above.
+            keep_as_earliest_frames: see above.
 
         """
-        if keep_last_exp and self.total_size > 0:
+        # TODO(breakds): Currently ``clear()`` is not handling
+        # ``self._keep_episodic_info`` properly. Therefore we temporaribly
+        # prohibit calling ``clear()`` when ``keep_episodic_info`` is enabled.
+        # This will be handled in a future PR and have _indexed_pos properly
+        # cleared.
+        assert not self._keep_episodic_info, (
+            'Currently it is prohibited to keep episodic info while also '
+            'calling clear().')
+
+        # First compute the number of experiences to keep after clearing for
+        # each of the environment.
+        num_to_keep = 0
+        # keep_last_exp requires keeping 1 extra experience from the end.
+        if keep_last_exp:
+            num_to_keep += 1
+        # keep_as_earliest_frames requires keeping num_earliest_frames_ignored
+        # extra experiences from the end.
+        if keep_as_earliest_frames:
+            num_to_keep += self._num_earliest_frames_ignored
+
+        # Do nothing if there is only so many experiences (per environment) in
+        # the buffer.
+        if num_to_keep >= self._current_size.max():
+            return
```

instead of silently return and potentially causing problem with RingBuffer, maybe we should just assert num_to_keep < self._current_size.min()?  Sorry for the delayed comment.